lib/teleterm/vnet/service.go

```diff
@@ -18,31 +18,258 @@ package service
+		return trace.BadParameter("missing DaemonService")
+	}
+
+	if c.ClientStore == nil {
+		return trace.BadParameter("missing ClientStore")
+	}
+
+	return nil
 }
 
 func (s *Service) Start(ctx context.Context, req *api.StartRequest) (*api.StartResponse, error) {
-	n := rand.Intn(10)
-	randomDelay := time.Duration(n) * 100 * time.Millisecond
-	time.Sleep(randomDelay + 400*time.Millisecond)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.isClosed {
+		return nil, trace.CompareFailed("VNet service has been closed")
+	}
+
+	if s.isRunning {
+		return &api.StartResponse{}, nil
+	}
+
+	longCtx, cancelLongCtx := context.WithCancelCause(context.Background())
+	s.cancel = cancelLongCtx
+	defer func() {
+		// If by the end of this RPC the service is not running, make sure to cancel the long context.
+		if !s.isRunning {
+			cancelLongCtx(nil)
+		}
+	}()
+
+	appProvider := &appProvider{
+		daemonService: s.cfg.DaemonService,
+		clientStore:   s.cfg.ClientStore,
+	}
+
+	manager, adminCommandErrCh, err := vnet.Setup(ctx, longCtx, appProvider)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.stopErrC = make(chan error, 1)
+
+	go func() {
+		err := vnet.Run(longCtx, cancelLongCtx, manager, adminCommandErrCh)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.ErrorContext(longCtx, "VNet closed with an error", "error", err)
+			s.stopErrC <- err
+		}
+		close(s.stopErrC)
+
+		// TODO(ravicious): Notify the Electron app about change of VNet state, but only if it's
+		// running. If it's not running, then the Start RPC has already failed and forwarded the error
+		// to the user.
+
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		s.isRunning = false
+		cancelLongCtx(nil)
+	}()
+
+	s.isRunning = true
 	return &api.StartResponse{}, nil
 }
 
+// Stop stops VNet and cleans up used resources. Blocks until VNet stops or ctx is canceled.
 func (s *Service) Stop(ctx context.Context, req *api.StopRequest) (*api.StopResponse, error) {
-	return &api.StopResponse{}, nil
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	errC := make(chan error)
+
+	go func() {
+		errC <- trace.Wrap(s.stopLocked())
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errC:
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return &api.StopResponse{}, nil
+	}
+
 }
 
-// Close stops the current VNet instance and prevents new instances from being started.
-//
+func (s *Service) stopLocked() error {
+	if s.isClosed {
+		return trace.CompareFailed("VNet service has been closed")
+	}
+
+	if !s.isRunning {
+		return nil
+	}
+
+	s.cancel(nil)
+	s.isRunning = false
+
+	return trace.Wrap(<-s.stopErrC)
+}
+
+// Close stops VNet service and prevents it from being started again. Blocks until VNet stops.
 // Intended for cleanup code when tsh daemon gets terminated.
 func (s *Service) Close() error {
-	return nil
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.stopLocked()
+	s.isClosed = true
+
+	return trace.Wrap(err)
+}
+
+type appProvider struct {
```

Comment:
Pretty much every `appProvider` method except `GetCachedClient` and `ReissueAppCert` was copied from [the tsh implementation](https://github.com/gravitational/teleport/blob/49918e2e3ac784244c4645fd535852bfcf70e94c/tool/tsh/common/vnet_common.go#L36-L42).

I thought about un-abstracting the methods that are the same between tsh and VNet, but idk, maybe it's fine the way it is. It should also make it easier to add tests in the future.